Contributing Guidelines (#5401)
aws-amplify/amplify-js: CONTRIBUTING.md

Opinionated implementations: There are many ways to interface with AWS Services. Certain service interactions are favored over others. For instance if sending and receiving JSON we would prefer an API Gateway endpoint to other mechanisms. Amplify will programmatically help optimize for cost and performance through library decisions.

Declarative actions: Amplify will provide you a reference to a generic client object and ability to perform common actions. “RegisterUser”, “Login”, “SendObject”, “UpdateObject”, “StreamData”. By default you should not need to worry about AWS Service specific API operations like putItem() with a unique hash or even HTTP verbs.

> Cascading service interactions: Certain actions in a declarative style can have overlapping or ambiguous AWS Service implementations. With an opinionated implementation we can decide which Services are "primary" and which are "secondary" depending on what is configured. For instance sending an image will prefer S3 over API Gateway..

Reviewer: ```suggestion
Cascading service interactions: Certain actions in a declarative style can have overlapping or ambiguous AWS Service implementations. With an opinionated implementation we can decide which Services are "primary" and which are "secondary" depending on what is configured. For instance sending an image will prefer S3 over API Gateway.
```

- Yarn build
- Run test suite
- Test in sample app using yarn linking
- Submit a PR

#### Build step:

```
$ yarn build --scope @aws-amplify/auth
```

#### Testing:

```
$ yarn run test --scope @aws-amplify/auth
$ yarn run test --scope aws-amplify-react
```

> Note: There is a commit hook that will run the tests prior to committing. Please make sure if you are going to provide a pull request to be sure you include unit tests with your functionality and that all tests are passing.

#### Test in a local sample app

**Yarn Linking**
The best way to develop locally and test is to link the individual package you’re working on and run lerna in watch mode.

Run watch mode while editing (auth for example):

```
$ npx lerna exec --scope @aws-amplify/auth yarn link
$ npx lerna exec --scope @aws-amplify/auth yarn build:esm:watch
```

Or run the whole library in watch mode if you are working on multiple packages

```
$ yarn build # Build the whole library
$ yarn link-all # Make all the packages available to link
$ yarn build:esm:watch # All packages are building ES6 modules in watch mode
```

In you sample project, you can now link specific packages

```
$ yarn link @aws-amplify/auth
```

>     These tests are only necessary if you’re looking to contribute a PR. If you’re just playing locally you don’t need them. However if you’re contributing a PR for anything other than bug fixes it would be best to validate that first because depending on the scope of the change.

Reviewer: ```suggestion
These tests are only necessary if you’re looking to contribute a PR. If you’re just playing locally you don’t need them. However if you’re contributing a PR for anything other than bug fixes it would be best to validate that first because depending on the scope of the change.
```

# Debugging:

Sometimes the issue can be solved by doing a clean and fresh build. To do this, make sure to remove your node modules and clean your packages. You can run `git clean -xdf` in the repository root to achieve this.

# Release

To give the bird's eye view of the release cycle,
- We follow semantic versioning for our releases
- Every merge into the `master` ends up as `unstable` package in the npm
- The core team will cut a release out to `stable` from `unstable` bi-weekly

## Finding contributions to work on

Looking at the existing issues is a great way to find something to contribute on. As our projects, by default, use the default GitHub issue labels (enhancement/bug/duplicate/help wanted/invalid/question/wontfix), looking at any ['help wanted'](https://github.com/aws-amplify/amplify-ios/labels/help%20wanted) or ['good first issue'](https://github.com/aws-amplify/amplify-ios/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22) is a great place to start.

You could also contribute by reporting bugs, reproduction of bugs with sample code, documentation and test improvements.

## Related Repositories

> The Amplify Framework runs on Android, iOS, and numerous JavaScript-based web platforms. T

Reviewer: Is a sentence missing here?

```suggestion
The Amplify Framework runs on Android, iOS, and numerous JavaScript-based web platforms.
```

3. [AWS Amplify CLI](https://github.com/aws-amplify/amplify-cli)

AWS Amplify plugins are built on top of the AWS SDKs. AWS SDKs are a
toolkit for interacting with AWS backend resources.

1. [AWS SDK for Android](https://github.com/aws-amplify/aws-sdk-android)
2. [AWS SDK for iOS](https://github.com/aws-amplify/aws-sdk-ios)
3. [AWS SDK for JavaScript](https://github.com/aws/aws-sdk-js)

## Code of Conduct

This project has adopted the [Amazon Open Source Code of Conduct](https://aws.github.io/code-of-conduct).
For more information see the [Code of Conduct FAQ](https://aws.github.io/code-of-conduct-faq) or contact
opensource-codeofconduct@amazon.com with any additional questions or comments.

## Security issue notifications

If you discover a potential security issue in this project we ask that you notify AWS/Amazon Security via our [vulnerability reporting page](http://aws.amazon.com/security/vulnerability-reporting/). Please do **not** create a public github issue.

## Licensing

See the [LICENSE](https://github.com/aws-amplify/amplify-ios/blob/master/LICENSE) file for our project's licensing. We will ask you to confirm the licensing of your contribution.

> We may ask you to sign a [Contributor License Agreement (CLA)](http://en.wikipedia.org/wiki/Contributor_License_Agreement) for larger changes..

Reviewer: ```suggestion
We may ask you to sign a [Contributor License Agreement (CLA)](http://en.wikipedia.org/wiki/Contributor_License_Agreement) for larger changes.
```

Reviewer 2: Is this necessary, given that PRs have:
> By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

(Admittedly, I've had to sign CLAs for other projects, but I like the frictionless PR flow for ours!)

$ yarn link @aws-amplify/auth
```

> These tests are only necessary if you’re looking to contribute a PR. If you’re just playing locally you don’t need them. However if you’re contributing a PR for anything other than bug fixes it would be best to validate that first because depending on the scope of the change.

Reviewer 3: ```suggestion
These tests are only necessary if you’re looking to contribute a Pull Request. If you’re just playing locally you don’t need them. However if you’re contributing a Pull Request for anything other than bug fixes it would be best to validate that first because depending on the scope of the change.
```
